File: LeetCode/538把二叉搜索树转换为累加树.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    private void calc(TreeNode root, int sum) {
       Stack<TreeNode> sta = new Stack<>();
       TreeNode cur = root;
       while (cur != null || !sta.isEmpty()) {
           if (cur != null) {
               sta.add(cur);
               cur = cur.right;
           } else {
               cur = sta.pop();
               sum += cur.val;
               cur.val = sum;
               cur = cur.left;
           }
       }
       return ;
    }

    public TreeNode convertBST(TreeNode root) {
        int sum = 0;
        calc(root, sum);
        return root;
    }
}

File: LeetCode/538把二叉搜索树转换为累加树.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func calc(root *TreeNode, sum *int) {
	if root == nil {
		return
	}
	calc(root.Right, sum)
	*sum += root.Val
	root.Val = *sum
	calc(root.Left, sum)
}

func convertBST(root *TreeNode) *TreeNode {
	sum := 0
	calc(root, &sum)
	return root
}